alica_engine/include/engine/model/Behaviour.h

```diff
@@ -27,54 +24,58 @@ class Behaviour : public AbstractPlan
 {
 public:
     Behaviour();
-    virtual ~Behaviour();
 
     std::string toString(std::string indent = "") const;
 
+    void setDeferring(int deferring);
```

Comment:
I don't think having public setters for each field is a elegant solution to building an object. It provides unnecessary set APIs publicly although these are only ever used for object building & secondly there is no single place to check if the object is built with all required things set. I think using the builder pattern here is possible or I would even just prefer to let the factory method access the private data members directly (since the factory method is anyway a creational pattern & the builder will have to do that anyway too). So friend here is probably not a bad thing imo

Reply:
Imo the two part construction obscures what is actually necessary to construct the class.  I suppos one question is: If all of the fields only exist to be set at construction time, but there are too many of them, why not just treat these objects as structs, make everything public, and then pass as const to the rest of the program?

Reply:
Couple of problems if we use plain structs I think - we tie the object representation to all parts of the program. Second, there are actually methods that do more than just set/get members in some model objects I think

Reply:
My only other idea without deeper refactoring is to create a struct used just to initialize.  Any other ideas?  I just want it to be clear what is used for construction and ideally make it hard to miss something

Reply:
but using struct as an argument not protects us against field initialization became missing, but not in result object, in struct member it could be

Reply:
yeah, but the constructor can verify if all required arguments are supplied & throw an exception if not

Reply:
for POD types it's barely possible. Non-pod types most of the time has a default ctor. Only if we will bring some heuristic which checks ranges for example...

Reply:
I'm not sure I follow your comment. My only intent was that if the model object is constructed at once using the constructor supplied with all the arguments it requires, then the constructor can verify if the arguments are valid & throw an exception if not. The way it is right now, the model object does not do any verification. Anyone can use any set method and change the object as they want

Reply:
I also prefer to either use a builder pattern or use structs to initialize model objects instead of adding setters which are only used during construction.